### PR TITLE
Fix GatedDeltaNet A_log dtype to prevent -inf under bfloat16 init

### LINKED ISSUE #47831: `Qwen3_5GatedDeltaNet` / `Qwen3_5MoeGatedDeltaNet` initialize `A_log` to `-inf` on some heads (bf16 models)
### System Info

- `transformers` version: 5.14.0
- Platform: Linux (glibc 2.34)
- Python version: 3.12.5
- PyTorch version (GPU?): 2.13.0+cu126 (yes, 8× NVIDIA H200)
- Model: Qwen3.5-MoE hybrid (40 layers = 30 GatedDeltaNet + 10 full attention, 256 experts), trained from scratch in bf16

Output of `transformers env` is consistent with the above.

### Who can help?

Text/hybrid (linear-attention) models — Qwen3.5 GatedDeltaNet (@ArthurZucker @Cyrilvallez @vasqu)

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

When a Qwen3.5 (`qwen3_5`) or Qwen3.5-MoE (`qwen3_5_moe`) model is instantiated in
bfloat16 (the normal case for `from_config(..., dtype=torch.bfloat16)` or
`from_pretrained(..., dtype=torch.bfloat16)` on a fresh config), a fraction of the
GatedDeltaNet heads get `A_log = -inf`, i.e. `A = exp(A_log) = 0`.

The cause is the decay-parameter initialization in
`Qwen3_5GatedDeltaNet.__init__` / `Qwen3_5MoeGatedDeltaNet.__init__`:

```python
A = torch.empty(self.num_v_heads).uniform_(0, 16)   # inherits model dtype (bf16)
self.A_log = nn.Parameter(torch.log(A))
```

`torch.empty(self.num_v_heads)` inherits the module's default dtype. Under a bf16
model that is **bfloat16**, where `uniform_(0, 16)` rounds a draw down to exactly
`0.0` with probability ≈ 0.4% per head. `torch.log(0) = -inf`, so that head's
decay `A = exp(A_log) = 0`.

Consequences for the affected head:
- `g = -A * softplus(a + dt_bias) = 0` -> the gated-delta recurrence **never
  decays/forgets** for that head (its state accumulates without bound).
- `dg/dA_log = -A * softplus(...) = 0` -> the gradient w.r.t. `A_log` is exactly 0,
  so the head is **permanently frozen at `-inf`** and can never recover.

Because the probability compounds over `num_value_heads × num_gdn_layers`, small
models are usually spared while large hybrids are reliably affected. On our 35B-A3B
config (32 value heads × 30 GDN layers ≈ 960 draws) we observe ~3–6 dead heads at
every initialization; an 8×8 tiny config shows none.

## Reproduction

```python
import torch

torch.manual_seed(0)
num_value_heads, num_gdn_layers = 32, 30

def count_dead(dtype):
    dead = 0
    for _ in range(num_gdn_layers):
        A = torch.empty(num_value_heads, dtype=dtype).uniform_(0, 16)  # current code
        A_log = torch.log(A.float()) if dtype == torch.float16 else torch.log(A)
        dead += int((~torch.isfinite(A_log)).sum())
    return dead

print("bfloat16 dead heads:", count_dead(torch.bfloat16))  # e.g. 6
print("float32  dead heads:", count_dead(torch.float32))   # 0
```

Or directly on a model:

```python
import torch
from transformers import AutoModelForCausalLM, AutoConfig

config = AutoConfig.from_pretrained("<a qwen3_5_moe hybrid config>")
model = AutoModelForCausalLM.from_config(config, dtype=torch.bfloat16)

for name, p in model.named_parameters():
    if name.endswith("A_log") and not torch.isfinite(p).all():
        print(name, "->", int((~torch.isfinite(p)).sum()), "dead head(s)")
```

### Expected behavior

`A_log` should be finite for every head at initialization, regardless of the model
dtype. No head should start with a zero (never-forgetting) decay gate.

## Suggested fix

Sample `A` in float32 before taking the log, matching the reference
implementations already in the repo. This is a 1-line change in each of
`Qwen3_5GatedDeltaNet.__init__` and `Qwen3_5MoeGatedDeltaNet.__init__`:

```diff
- A = torch.empty(self.num_v_heads).uniform_(0, 16)
+ A = torch.empty(self.num_v_heads, dtype=torch.float32).uniform_(0, 16)
  self.A_log = nn.Parameter(torch.log(A))
```

This aligns with existing models in the codebase, which already build `A` in
float32:
- `Mamba2Mixer` (`modeling_mamba2.py`)
- `Qwen3NextGatedDeltaNet` (`modeling_qwen3_next.py`)

so the fix keeps Qwen3.5 consistent with the rest of the library rather than
diverging. The `dt_bias`/`A_log` parameters are also intended to be decay-free;
keeping them float32 additionally avoids losing precision on the decay term.

I'm happy to open the PR with this change if it looks good to the maintainers.

### PR DESCRIPTION
**Pull request by Nkluge-correa**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47944)

# What does this PR do?

`Qwen3_5GatedDeltaNet`, `Qwen3_5MoeGatedDeltaNet`, and `Qwen3NextGatedDeltaNet` sample `A_log` via:

```python
A = torch.empty(self.num_v_heads).uniform_(0, 16)
self.A_log = nn.Parameter(torch.log(A))
```

and re-sample it the same way in each model's `_init_weights`. When the model is initialized under a `bfloat16` default dtype (e.g., `.from_config(config, dtype=torch.bfloat16)`), `uniform_(0, 16)` can round a sampled value down to `0`, and `log(0) == -inf`. That `-inf` gets baked into `A_log`, permanently zeroing that head's decay term and freezing its gradient for the rest of training.

This PR samples `A` in `float32` explicitly (independent of the ambient default dtype) before taking the log, in both the module constructor and `_init_weights`, for all three affected models mentioned above.

It also adds a regression test (`test_gated_delta_net_a_log_finite_under_bfloat16_init`) for each of the three affected models asserting `A_log` stays `float32` and finite when the model is constructed with `dtype=torch.bfloat16`. I ran the full test suite for all three models locally with no regressions:

- `tests/models/qwen3_5/test_modeling_qwen3_5.py`: 292 passed, 232 skipped
- `tests/models/qwen3_5_moe/test_modeling_qwen3_5_moe.py`: 296 passed, 220 skipped
- `tests/models/qwen3_next/test_modeling_qwen3_next.py`: 148 passed, 145 skipped

Fixes #47831

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. These often are low-quality, or fix extremely minor issues that occur rarely or never in practice.
As a result, we're instituting a rule that **first-time contributors should not use code agents to submit PRs or issues**.
We'd also ask autonomous "OpenClaw"-like agents not to open any PRs or issues.

Issues/PRs from first-time contributors that violate this rule will probably just be closed without review, and we
might block you, especially if you open more than one or appear to be deliberately ignoring this. We especially do not
want new contributors to jump in on random issues to contribute an agent-written fix. This creates lots of noise
for reviewers and other users and will almost certainly get you blocked.

For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [x] (First-time contributors only): I confirm that this PR description and code is not written by an LLM or code agent

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/huggingface/transformers/issues/47831
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@Rocketknight1

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by Rocketknight1 — COMMENTED** [review]
Hey! The bug seems real but I'm unsure about the fix - it involves a lot of dtype casting, and I think it might cause fp32 params to leak out into the model. Why not just have a simpler fix: Bump the lower bound of the `uniform_` call to `0.01` or something? This preserves init while avoiding 0 values, and simplifies everything drastically. You could also remove the test as well after this, since a small init value change like that shouldn't require testing. A single comment explaining that we don't want 0 values should be enough to stop regression

**Comment by Nkluge-correa** [comment]
Sure, sounds much simpler to be honest. Switched to sampling `A` with `uniform_(0.01, 16)` (in both the constructor and `_init_weights`) instead of the float32 cast. Also added the one-line comment explaining why the lower bound is nonzero. Also removed the regression test. Ran the three affected test suites again locally with no regressions. Thanks for the suggestion!

**Review by Rocketknight1 — APPROVED** [review]
Yes, looks good now, thank you!
